packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+    /// ptr += (userLen OR 20) + 1
+    /// [ptr(prodcutLen), ptr+1:productLen(product registry # OR 20 byte address if prdoctLen == 0)] => address product
+    /// ptr += (prodcutLen OR 20) + 1
+    /// [ptr(amountLen), ptr:amountLen] => uint256 amount 
+
+
+    // ARGUMENT TYPE DECODING //
+
+    function decodeAddressAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, address addr2, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return (addr1, addr2, amount, ptr);
+    }
+
+    function decodeAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return(addr1, UFixed18(amount), ptr);
+    }
+
+    function decodeProductPrograms(bytes calldata input, uint ptr) private returns(address product, uint256[] memory programs, uint) {
+        (product, ptr) = decodeAddress(input, ptr);
+        (programs, ptr) = decodeUintArray(input, ptr);
+
+        return(product, programs, ptr);
+    }
+
+    function decodeAddressAddress(bytes calldata input, uint ptr) private returns(address addr1, address addr2, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+
+        return (addr1, addr2, ptr);
+    }
+
+
+    // INDIVIDUAL TYPE DECODING //
+
+     function decodeAmountUFixed18(bytes calldata input, uint ptr) private returns (UFixed18 result, uint) {
+        uint temp;
+        (temp, ptr) = decodeUint(input, ptr);
+
+        return(UFixed18.wrap(temp), ptr);
+    }
+
+    function decodeUint(bytes calldata input, uint ptr) private returns (uint result, uint) {
+        uint8 len = toUint8(input[ptr:ptr+1]);
+        ++ptr;
+
+        result = bytesToUint(input[ptr:ptr+len]);
+        ptr += len;
+
+        return (result, ptr);
+    }
+
+ 
+
+    // function decodeUFixed18Array(bytes calldata input, uint ptr) private returns (UFixed18[] memory, uint) {
+    //     uint8 arrayLen = toUint8(input[ptr:ptr+1]);
+    //     ++ptr;
+
+    //     UFixed18[] memory result = new UFixed18[](arrayLen);
+    //     uint count = 0;
+    //     for(;count < arrayLen;) {
+    //         UFixed18 currUint;
+
+    //         (currUint, ptr) = decodeAmountUFixed18(input, ptr);
+            
+    //         result[count] = currUint;
+
+    //         ++count;
+    //     }
+
+    //     return (result, ptr);
+    // }
+
+
+    function decodeUintArray(bytes calldata input, uint ptr) private returns (uint256[] memory, uint) {
+        uint8 arrayLen = toUint8(input[ptr:ptr+1]);
+        ++ptr;
+
+        uint256[] memory result = new uint256[](arrayLen);
+
+        uint count = 0;
```

Comment:
why is this on its own line instead of inside the for definition?

Reply:
bad habit saves 2 gas not worth it